File: study/ts/script/scene/view/petMain/PetMainAvatorsNode.ts
const { ccclass, property } = cc._decorator;

import CommonPetMainAvatar from '../../../ui/component/CommonPetMainAvatar'
import { G_UserData, G_SignalManager, G_Prompt } from '../../../init';
import { Path } from '../../../utils/Path';
import PetConst from '../../../const/PetConst';
import { SignalConst } from '../../../const/SignalConst';
import { handler } from '../../../utils/handler';
import { FunctionConst } from '../../../const/FunctionConst';
import { RedPointHelper } from '../../../data/RedPointHelper';
import { LogicCheckHelper } from '../../../utils/LogicCheckHelper';
import CircleScroll from '../../../ui/CircleScroll';

@ccclass
export default class PetMainAvatorsNode extends CircleScroll {
    @property({
        type: cc.Node,
        visible: true
    })
    _resourceNode: cc.Node = null;
    @property({
        type: CommonPetMainAvatar,
        visible: true
    })
    _avatar5: CommonPetMainAvatar = null;

    @property({
        type: CommonPetMainAvatar,
        visible: true
    })
    _avatar6: CommonPetMainAvatar = null;

    @property({
        type: CommonPetMainAvatar,
        visible: true
    })
    _avatar4: CommonPetMainAvatar = null;

    @property({
        type: CommonPetMainAvatar,
        visible: true
    })
    _avatar3: CommonPetMainAvatar = null;

    @property({
        type: CommonPetMainAvatar,
        visible: true
    })
    _avatar2: CommonPetMainAvatar = null;

    @property({
        type: CommonPetMainAvatar,
        visible: true
    })
    _avatar1: CommonPetMainAvatar = null;

    @property({
        type: CommonPetMainAvatar,
        visible: true
    })
    _avatar7: CommonPetMainAvatar = null;
    _startIndex: any;
    _parentView: any;
    _signalPetLevelUpdate;
    _signalChangePetFormation;
    _signalRedPointUpdate;
    _signalPetLimitUp: import("f:/mingjiangzhuan/main/assets/resources/script/utils/event/Slot").Slot;

    ctorSlef(size, angles, startIndex, parentView, angleOffset, circle, scaleRange) {
        super.ctor(size, angles, startIndex, angleOffset, circle, scaleRange);
        this._resourceNode = null;
        this._startIndex = startIndex;
        this._parentView = parentView;
        this.node.name = "PetMainAvatorsNode";
    }
    _preCreateAvatar() {
        var [showNum] = G_UserData.getPet().getShowPetNum();
        let createAvatar = function (index) {
            var petAvatar = this['_avatar' + index];
            petAvatar.updateImageBk(Path.getPet('pet_taizi' + index));
            petAvatar.node.name = ('_avatar' + index);
            this['_avatar' + index] = petAvatar;
            return petAvatar;
        }.bind(this);
        if (showNum <= PetConst.SCROLL_AVATART_NUM) {
            for (var i = 1; i <= showNum; i++) {
                var petInfo = PetConst['PET_INFO' + showNum];
                var position = petInfo[i].position;
                var scale = petInfo[i].scale;
                var zorder = petInfo[i].zorder;
                var imageScale = petInfo[i].imageScale;
                var petAvatar = createAvatar(i);
                petAvatar.node.setPosition(position);
                petAvatar.setAvatarScale(scale);
                petAvatar.setImageScale(imageScale);
                petAvatar.node.zIndex = (zorder);
                this.addMidLayer(petAvatar.node);
            }
        } else {
            for (var i = 1; i <= showNum; i++) {
                var petAvatar = createAvatar(i);
                petAvatar.setAvatarScale(PetConst.SCROLL_AVATAR_SCALE);
                this.addNode(petAvatar.node, i);
            }
        }
        for (var i = showNum + 1; i <= 7; i++) {
            this['_avatar' + i].node.active = false;
            this['_avatar' + i] = null;
        }
    }
    updateAll() {
        this._updatePetAvatar();
    }
    start() {
        this._preCreateAvatar();
        this._playNodeEffect();
        this._signalPetLevelUpdate = G_SignalManager.add(SignalConst.EVENT_PET_LEVEL_UP_SUCCESS, handler(this, this._onEventPetLevelUpdate));
        this._signalChangePetFormation = G_SignalManager.add(SignalConst.EVENT_PET_ON_TEAM_SUCCESS, handler(this, this._onEventUserPetChange));
        this._signalRedPointUpdate = G_SignalManager.add(SignalConst.EVENT_RED_POINT_UPDATE, handler(this, this._onEventRedPointUpdate));
        this._signalPetLimitUp = G_SignalManager.add(SignalConst.EVENT_PET_LIMITUP_SUCCESS, handler(this, this._onEventUserPetChange));
        this.updateAll();
        G_UserData.getAttr().recordPower();
        G_SignalManager.dispatch(SignalConst.EVENT_TUTORIAL_STEP, "PetMainAvatorsNode");
    }
    onDestroy() {
        this._signalPetLevelUpdate.remove();
        this._signalPetLevelUpdate = null;
        this._signalChangePetFormation.remove();
        this._signalChangePetFormation = null;
        this._signalRedPointUpdate.remove();
        this._signalRedPointUpdate = null;
        this._signalPetLimitUp.remove();
        this._signalPetLimitUp = null;
    }
    _onClickPetAvatar(mainAvatar) {
    }
    getAvatarIndex(funcId) {
        var funcIdList = this._getFuncList();
        for (var i in funcIdList) {
            var value = funcIdList[i];
            if (value == funcId) {
                return parseFloat(i);
            }
        }
        return 1;
    }
    _getFuncList() {
        var funcIdList = {};
        var [showNum] = G_UserData.getPet().getShowPetNum();
        if (showNum <= 5) {
            funcIdList = {
                [1]: FunctionConst.FUNC_PET_HELP_SLOT1,
                [2]: FunctionConst.FUNC_PET_HELP_SLOT2,
                [3]: FunctionConst.FUNC_PET_HELP_SLOT3,
                [4]: FunctionConst.FUNC_PET_HELP_SLOT4,
                [5]: FunctionConst.FUNC_PET_HELP_SLOT5
            };
        } else {
            funcIdList = {
                [1]: FunctionConst.FUNC_PET_HELP_SLOT1,
                [2]: FunctionConst.FUNC_PET_HELP_SLOT2,
                [3]: FunctionConst.FUNC_PET_HELP_SLOT3,
                [4]: FunctionConst.FUNC_PET_HELP_SLOT4,
                [5]: FunctionConst.FUNC_PET_HELP_SLOT5,
                [6]: FunctionConst.FUNC_PET_HELP_SLOT6,
                [7]: FunctionConst.FUNC_PET_HELP_SLOT7
            };
        }
        return funcIdList;
    }
    getFuncIdByIndex(index) {
        var funcIdList = this._getFuncList();
        return funcIdList[index];
    }
    _playNodeEffect() {
        for (var i = 4; i <= 6; i++) {
            var avatarNode = this['_avatar' + i];
            if (avatarNode) {
                avatarNode.playEffect('effect_shenshou_taizi' + i);
            }
        }
    }
    _updatePetAvatar() {
        var petIdList = G_UserData.getTeam().getPetIdsInHelpWithZero();
        var startFuncId = FunctionConst.FUNC_PET_HELP_SLOT1;
        for (var index = 1; index <= 7; index++) {
            var avatarNode = this['_avatar' + index];
            if (avatarNode && avatarNode.node.active) {
                var funcTeamSoltId = this.getFuncIdByIndex(index);
                avatarNode.setFuncId(funcTeamSoltId);
            }
        }
        let onlyUnlockOne = function () {
            var unLockOne = 0;
            var lockLevel = 0;
            var retCanMove = true;
            for (var index = FunctionConst.FUNC_PET_HELP_SLOT1; index <= FunctionConst.FUNC_PET_HELP_SLOT7; index++) {
                var avatarIndex = this.getAvatarIndex(index);
                var avatarNode = this['_avatar' + avatarIndex];
                if (avatarNode) {
                    var funcTeamSoltId = this.getFuncIdByIndex(avatarIndex);
                    var petId = petIdList[funcTeamSoltId - FunctionConst.FUNC_PET_HELP_SLOT1];
                    var [isOpen] = LogicCheckHelper.funcIsOpened(funcTeamSoltId);
                    var isShow = LogicCheckHelper.funcIsShow(funcTeamSoltId);
                    avatarNode.node.active = (isShow);
                    if (isShow == false) {
                        retCanMove = false;
                    }
                    avatarNode.setLock(!isOpen);
                    avatarNode.setAdd(isOpen);
                    if (isOpen == false && lockLevel == 0 && petId == 0) {
                        avatarNode.showOpenLevel(true);
                        lockLevel = lockLevel + 1;
                    }
                }
            }
            return retCanMove;
        }.bind(this);
        this.setMoveEnable(false);
        var moveEnable = onlyUnlockOne();
        this.setMoveEnable(moveEnable);
        for (var i in petIdList) {
            var value = petIdList[i];
            var petUnit = G_UserData.getPet().getUnitDataWithId(value);
            var avatarIndex = this.getAvatarIndex(parseFloat(i) + FunctionConst.FUNC_PET_HELP_SLOT1);
            var funcTeamSoltId = this.getFuncIdByIndex(avatarIndex);
            var avatarNode = this['_avatar' + avatarIndex];
            if (petUnit && avatarNode) {
                avatarNode.updateUI(petUnit.getBase_id(), handler(this, this._onClickPetAvatar));
                avatarNode.updatePetName(petUnit.getBase_id(), petUnit.getStar(), petUnit.getLevel());
                avatarNode.setPetIndex(funcTeamSoltId - FunctionConst.FUNC_PET_HELP_SLOT1 + 1);
            }
        }
        this._updatePetAvatarRedPoint();
    }
    _updateRedPointByFuncId(funcId, param) {
        if (funcId && funcId > 0) {
            if (this._isInPetAvatarFuncList(funcId)) {
                this._updatePetAvatarRedPoint(funcId, param);
            }
        }
    }
    _onEventUserDataUpdate(_, param) {
        this._updatePetAvatar();
    }
    _onEventPetLevelUpdate(_, param) {
        this._updatePetAvatar();
    }
    _onEventUserPetChange(_, param) {
        G_UserData.getAttr().recordPower();
        this._updatePetAvatar();
        G_Prompt.playTotalPowerSummary();
    }
    _onEventOfficialLevelUp(_, param) {
        this._updatePetAvatar();
    }
    _onEventRedPointUpdate(id, funcId, param) {
        this._updateRedPointByFuncId(funcId, param);
    }
    _isInPetAvatarFuncList(funcId) {
        var list = {};
        for (var i in list) {
            var id = list[i];
            if (id == funcId) {
                return true;
            }
        }
        return false;
    }
    _updatePetAvatarRedPoint(funcId?, param?) {
        function checkPetUpgrade(petId) {
            var petUnitData = G_UserData.getPet().getUnitDataWithId(petId);
            if (petUnitData == null) {
                return false;
            }
            var reach = RedPointHelper.isModuleReach(FunctionConst.FUNC_PET_TRAIN_TYPE1, petUnitData);
            return reach;
        }
        function checkPetBreak(petId) {
            var petUnitData = G_UserData.getPet().getUnitDataWithId(petId);
            if (petUnitData == null) {
                return false;
            }
            var reach = RedPointHelper.isModuleReach(FunctionConst.FUNC_PET_TRAIN_TYPE2, petUnitData);
            return reach;
        }
        function checkPetLimit(petId) {
            var petUnitData = G_UserData.getPet().getUnitDataWithId(petId);
            var reach = RedPointHelper.isModuleReach(FunctionConst.FUNC_PET_TRAIN_TYPE3, petUnitData);
            return reach;
        }
        var checkFuncs = {
            [FunctionConst.FUNC_PET_TRAIN_TYPE1]: checkPetUpgrade,
            [FunctionConst.FUNC_PET_TRAIN_TYPE2]: checkPetBreak,
            [FunctionConst.FUNC_PET_TRAIN_TYPE3]: checkPetLimit
        };
        var redPointFuncId = [
            FunctionConst.FUNC_PET_TRAIN_TYPE1,
            FunctionConst.FUNC_PET_TRAIN_TYPE2,
            FunctionConst.FUNC_PET_TRAIN_TYPE3
        ];
        var petIdList = G_UserData.getTeam().getPetIdsInHelpWithZero();
        for (var index = FunctionConst.FUNC_PET_HELP_SLOT1; index <= FunctionConst.FUNC_PET_HELP_SLOT7; index++) {
            var avatarIndex = this.getAvatarIndex(index);
            var avatarNode = this['_avatar' + avatarIndex];
            if (avatarNode) {
                var funcTeamSoltId = this.getFuncIdByIndex(avatarIndex);
                var petId = petIdList[funcTeamSoltId - FunctionConst.FUNC_PET_HELP_SLOT1];
                if (petId > 0) {
                    var reachRedPoint = false;
                    for (var j in redPointFuncId) {
                        funcId = redPointFuncId[j];
                        var func = checkFuncs[funcId];
                        var reach = func(petId);
                        if (reach) {
                            reachRedPoint = true;
                            break;
                        }
                    }
                    avatarNode.showRedPoint(reachRedPoint);
                }
            }
        }
    }

}

File: study/lua/src/app/scene/view/petMain/PetMainAvatorsNode.lua

local CircleScroll = require("app.ui.CircleScroll")
local PetMainAvatorsNode = class("PetMainAvatorsNode", CircleScroll)
local FunctionConst = require("app.const.FunctionConst")
local LogicCheckHelper = require("app.utils.LogicCheckHelper")
local RedPointHelper = require("app.data.RedPointHelper")
local DataConst = require("app.const.DataConst")
local AvatarDataHelper = require("app.utils.data.AvatarDataHelper")
local PetConst = require("app.const.PetConst")
function PetMainAvatorsNode:ctor(size, angles, startIndex, parentView, angleOffset, circle, scaleRange)
    PetMainAvatorsNode.super.ctor(self, size, angles, startIndex,angleOffset,circle,scaleRange)
    self._resourceNode = nil
    self._startIndex = startIndex
	self._parentView = parentView

    self:setName("PetMainAvatorsNode")
    self:onCreate()
	--PetMainAvatorsNode.super.ctor(self, resource)
end

--预创建移动小岛
function PetMainAvatorsNode:_preCreateAvatar( ... )
    -- body
    local CSHelper  = require("yoka.utils.CSHelper")
    local PetConst = require("app.const.PetConst")
    local showNum = G_UserData:getPet():getShowPetNum()

    local function createAvatar(index)
        -- body
        local petAvatar = CSHelper.loadResourceNode(Path.getCSB("CommonPetMainAvatar", "common"))    
        petAvatar:updateImageBk(Path.getPet("pet_taizi"..index))
       
        petAvatar:setName("_avatar"..index)
        self["_avatar"..index] = petAvatar
        return petAvatar
    end


    if showNum <= PetConst.SCROLL_AVATART_NUM then
        for i= 1, showNum do
            local petInfo = PetConst["PET_INFO"..showNum]
            local position = petInfo[i].position
            local scale = petInfo[i].scale
            local zorder = petInfo[i].zorder
            local imageScale = petInfo[i].imageScale

            local petAvatar = createAvatar(i)
            petAvatar:setPosition(position)
            petAvatar:setAvatarScale(scale)
            petAvatar:setImageScale(imageScale)
            petAvatar:setLocalZOrder(zorder)
            self:addMidLayer(petAvatar)
        end
    else
        for i= 1, showNum do
            local petAvatar = createAvatar(i)
            petAvatar:setAvatarScale(PetConst.SCROLL_AVATAR_SCALE)
            self:addNode(petAvatar, i)
        end
    end
end

function PetMainAvatorsNode:onCreate()
    local CSHelper  = require("yoka.utils.CSHelper")

    local PetConst = require("app.const.PetConst")

    self:_preCreateAvatar()

    --[[
    local petAvatar7 = CSHelper.loadResourceNode(Path.getCSB("CommonPetMainAvatar", "common"))
    petAvatar7:setName("_avatar7")
    self["_avatar7"] = petAvatar7

    petAvatar7:setScale(1.0)
    petAvatar7:updateScale(1.0)
    petAvatar7:setShadowScale(2.7) --神兽影子放大
    petAvatar7:setAvatarScale(PetConst.MID_AVATAR_SCALE)
    
    petAvatar7:updateImageBk(Path.getPet("pet_taizi7"))
    petAvatar7:setPosition(PetConst.MIDDLE_POSITION)
    self:addMidLayer(petAvatar7)
    --
	for i = 1, 6 do
        local petAvatar = CSHelper.loadResourceNode(Path.getCSB("CommonPetMainAvatar", "common"))
        self:addNode(petAvatar, i)

        petAvatar:setName("_avatar"..i)
        petAvatar:updateImageBk(Path.getPet("pet_taizi"..i))
        petAvatar:setAvatarScale(PetConst.SCROLL_AVATAR_SCALE)
        
        self["_avatar"..i] = petAvatar
	end
    ]]
     self:_playNodeEffect()

end

function PetMainAvatorsNode:updateAll()
    logWarn("PetMainAvatorsNode:updateAll")
    self:_updatePetAvatar()
end

function PetMainAvatorsNode:onEnter()
    self._signalPetLevelUpdate     = G_SignalManager:add(SignalConst.EVENT_PET_LEVEL_UP_SUCCESS, handler(self, self._onEventPetLevelUpdate))
	self._signalChangePetFormation = G_SignalManager:add(SignalConst.EVENT_PET_ON_TEAM_SUCCESS, handler(self, self._onEventUserPetChange))
    self._signalRedPointUpdate      = G_SignalManager:add(SignalConst.EVENT_RED_POINT_UPDATE, handler(self,self._onEventRedPointUpdate))


    self:updateAll()

    G_UserData:getAttr():recordPower()
	--抛出新手事件出新手事件
    G_SignalManager:dispatch(SignalConst.EVENT_TUTORIAL_STEP, self.__cname)
end

function PetMainAvatorsNode:onExit()

    self._signalPetLevelUpdate:remove()
    self._signalPetLevelUpdate =nil

    self._signalChangePetFormation:remove()
    self._signalChangePetFormation =nil
  
    self._signalRedPointUpdate:remove()
    self._signalRedPointUpdate = nil

end

function PetMainAvatorsNode:_onClickPetAvatar(mainAvatar)

end


function PetMainAvatorsNode:getAvatarIndex(funcId)
    local funcIdList = self:_getFuncList()
    for i, value in ipairs(funcIdList) do
        if value == funcId then
            return i
        end
    end

    return 1
    -- body
end

function PetMainAvatorsNode:_getFuncList()
    local funcIdList = {}
     local showNum = G_UserData:getPet():getShowPetNum()
    if showNum <= 5 then
        funcIdList = {
            [1] = FunctionConst.FUNC_PET_HELP_SLOT1,
            [2] = FunctionConst.FUNC_PET_HELP_SLOT2,
            [3] = FunctionConst.FUNC_PET_HELP_SLOT3,
            [4] = FunctionConst.FUNC_PET_HELP_SLOT4,
            [5] = FunctionConst.FUNC_PET_HELP_SLOT5,
        }
    else
        funcIdList = {
            [1] = FunctionConst.FUNC_PET_HELP_SLOT1,
            [2] = FunctionConst.FUNC_PET_HELP_SLOT2,
            [3] = FunctionConst.FUNC_PET_HELP_SLOT3,
            [4] = FunctionConst.FUNC_PET_HELP_SLOT4,
            [5] = FunctionConst.FUNC_PET_HELP_SLOT5,
            [6] = FunctionConst.FUNC_PET_HELP_SLOT6,
            [7] = FunctionConst.FUNC_PET_HELP_SLOT7,
        }
    end
    return funcIdList
end

function PetMainAvatorsNode:getFuncIdByIndex(index)

    local funcIdList = self:_getFuncList()
    return funcIdList[index]
    -- body
end
function PetMainAvatorsNode:_playNodeEffect( ... )
    -- body
    for i =4 , 6 do
        local avatarNode = self["_avatar"..i]
        if avatarNode then
            avatarNode:playEffect("effect_shenshou_taizi"..i)
        end
    end
end
--更新上阵武将列表
function PetMainAvatorsNode:_updatePetAvatar()
    local petIdList = G_UserData:getTeam():getPetIdsInHelpWithZero()
    local startFuncId = FunctionConst.FUNC_PET_HELP_SLOT1

    local LogicCheckHelper = require("app.utils.LogicCheckHelper")

    for index= 1, 7 do
        local avatarNode = self["_avatar"..index]
        if avatarNode then
            local funcTeamSoltId= self:getFuncIdByIndex(index)
            avatarNode:setFuncId(funcTeamSoltId)
        end
    end
  
    --只解锁一个
    local function onlyUnlockOne()
        local unLockOne = 0
        local lockLevel = 0
        local retCanMove = true
        for index= FunctionConst.FUNC_PET_HELP_SLOT1, FunctionConst.FUNC_PET_HELP_SLOT7 do
            local avatarIndex = self:getAvatarIndex(index)
            local avatarNode = self["_avatar"..avatarIndex]
            if avatarNode then
                local funcTeamSoltId = self:getFuncIdByIndex(avatarIndex)
                local petId = petIdList[funcTeamSoltId - FunctionConst.FUNC_PET_HELP_SLOT1 + 1]
                local isOpen = LogicCheckHelper.funcIsOpened( funcTeamSoltId )
                local isShow = LogicCheckHelper.funcIsShow( funcTeamSoltId )
                avatarNode:setVisible(isShow)
                if isShow == false then
                    retCanMove = false
                end
                avatarNode:setLock(not isOpen)
                avatarNode:setAdd(isOpen)
              
                if isOpen == false and lockLevel == 0 and petId == 0 then
                    logWarn("avatarNode:showOpenLevel(true)")
                    avatarNode:showOpenLevel(true)
                    lockLevel = lockLevel + 1
                end
                
            end
        end
        return retCanMove
    end
    self:setMoveEnable(false)
    local moveEnable = onlyUnlockOne()
    self:setMoveEnable(moveEnable)
    
    for i, value in ipairs(petIdList) do
        local petUnit = G_UserData:getPet():getUnitDataWithId(value)
        local avatarIndex = self:getAvatarIndex(i + FunctionConst.FUNC_PET_HELP_SLOT1 -1)
        local funcTeamSoltId = self:getFuncIdByIndex(avatarIndex)
        local avatarNode = self["_avatar"..avatarIndex]
        if petUnit and avatarNode then
            avatarNode:updateUI(petUnit:getBase_id(), handler(self,self._onClickPetAvatar))
            avatarNode:updatePetName(petUnit:getBase_id(), petUnit:getStar(),petUnit:getLevel())
            avatarNode:setPetIndex(funcTeamSoltId - FunctionConst.FUNC_PET_HELP_SLOT1 + 1)
        end
    end
    self:_updatePetAvatarRedPoint()
end


function PetMainAvatorsNode:_updateRedPointByFuncId(funcId, param)
    if funcId and funcId > 0 then
        if self:_isInPetAvatarFuncList(funcId) then
            self:_updatePetAvatarRedPoint(funcId, param)
        end
    end
end


-- user数据更新
function PetMainAvatorsNode:_onEventUserDataUpdate(_, param)
    self:_updatePetAvatar()--改名
end

-- 角色升级，刷新按钮状态
function PetMainAvatorsNode:_onEventPetLevelUpdate(_, param)
end

-- 角色升级，刷新按钮状态
function PetMainAvatorsNode:_onEventUserPetChange(_, param)
    G_UserData:getAttr():recordPower()
    self:_updatePetAvatar()

    G_Prompt:playTotalPowerSummary()
end

function PetMainAvatorsNode:_onEventOfficialLevelUp(_, param)
    self:_updatePetAvatar()
end

--小红点刷新
function PetMainAvatorsNode:_onEventRedPointUpdate(id, funcId, param)
   self:_updateRedPointByFuncId(funcId, param)
end

--是否在Avatar红点检测列表中
function PetMainAvatorsNode:_isInPetAvatarFuncList(funcId)
    local list = {
    }
    for i, id in ipairs(list) do
        if id == funcId then
            return true
        end
    end
    return false
end


function PetMainAvatorsNode:_updatePetAvatarRedPoint(funcId, param)
   local function checkPetUpgrade(petId)
        local petUnitData = G_UserData:getPet():getUnitDataWithId(petId)
        if petUnitData == nil then
            return false
        end
        local reach = RedPointHelper.isModuleReach(FunctionConst.FUNC_PET_TRAIN_TYPE1, petUnitData)
        return reach
    end

    local function checkPetBreak(petId)
        local petUnitData = G_UserData:getPet():getUnitDataWithId(petId)
        if petUnitData == nil then
            return false
        end
        local reach = RedPointHelper.isModuleReach(FunctionConst.FUNC_PET_TRAIN_TYPE2, petUnitData)
        return reach
    end

    local function checkPetLimit(petId)
		local petUnitData = G_UserData:getPet():getUnitDataWithId(petId)
		local reach = RedPointHelper.isModuleReach(FunctionConst.FUNC_PET_TRAIN_TYPE3, petUnitData)
		return reach
	end

     local checkFuncs = {
        [FunctionConst.FUNC_PET_TRAIN_TYPE1] = checkPetUpgrade,
        [FunctionConst.FUNC_PET_TRAIN_TYPE2] = checkPetBreak,
        [FunctionConst.FUNC_PET_TRAIN_TYPE3] = checkPetLimit,
     }

     --红点相关的
    local redPointFuncId = {
        FunctionConst.FUNC_PET_TRAIN_TYPE1,
        FunctionConst.FUNC_PET_TRAIN_TYPE2,
        FunctionConst.FUNC_PET_TRAIN_TYPE3,
    }

    local petIdList = G_UserData:getTeam():getPetIdsInHelpWithZero()
    --护佑
    for index= FunctionConst.FUNC_PET_HELP_SLOT1, FunctionConst.FUNC_PET_HELP_SLOT7 do
        local avatarIndex = self:getAvatarIndex(index)
        local avatarNode = self["_avatar"..avatarIndex]
        if avatarNode then
            local funcTeamSoltId = self:getFuncIdByIndex(avatarIndex)
            local petId = petIdList[funcTeamSoltId - FunctionConst.FUNC_PET_HELP_SLOT1 + 1]
            if petId > 0 then
                local reachRedPoint = false
                for j, funcId in ipairs(redPointFuncId) do
                    local func = checkFuncs[funcId]
                    local reach = func(petId)
                    if reach then
                        reachRedPoint = true
                        break
                    end
                end
                avatarNode:showRedPoint(reachRedPoint)
            end
        end
    end

end


return PetMainAvatorsNode
